### ISSUE TITLE
[Bug] learning_to_remember_rare_events: shape mismatch when calculating teacher_loss

### ISSUE THREAD

**Issue by kasrahbar** [tensorflow/models#5306]
Hello,

It looks like there is bug when calculating teacher_loss because the shapes of neg_teacher_vals and teacher_vals don't match. 

https://github.com/tensorflow/models/blob/34beb7adfcd2f394e09acfecf05fdc0bdb8143c5/research/learning_to_remember_rare_events/memory.py#L216

The shape of teacher_vals is [batch_size] whereas neg_teacher_vals is [batch_size, 1]. This results in teacher_loss having a shape of [batch_size, batch_size], when it should really have a shape of [batch_size]. This bug is later masked by tf.reduce_mean(teacher_loss) at line 260 which returns a scalar.

This bug can be fixed by simply changing line 207 to
teacher_vals *= tf.expand_dims( 1 - tf.to_float(tf.equal(0.0, tf.reduce_sum(teacher_hints, 1))), 1)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by kasrahbar** [comment]
Thanks for the response. I updated the original post.

**Comment by karmel** [comment]
@lukaszkaiser , can you take a look at this shape mismatch for learning_to_remember?

**Comment by lukaszkaiser** [comment]
Sounds like you're right! Did you re-run the Omniglot experiments to make sure they still work after the correction?

**Comment by kasrahbar** [comment]
Rerunning Omniglot after fixing the bug adversely affects performance... I haven't found the reason yet for why that might be. Any ideas?

Before fix:
INFO:root:episode batch 500, avg train loss 0.105371
INFO:root:validation overall accuracy 0.758000
INFO:root:0-shot: 0.064, 1-shot: 0.836, 2-shot: 0.896, 3-shot: 0.876, 4-shot: 0.936, 5-shot: 0.940,

After fix:
INFO:root:episode batch 500, avg train loss 0.135186
INFO:root:validation overall accuracy 0.587333
INFO:root:0-shot: 0.032, 1-shot: 0.516, 2-shot: 0.712, 3-shot: 0.704, 4-shot: 0.780, 5-shot: 0.780,

**Comment by kasrahbar** [comment]
I was a bit quick to jump to conclusions in my previous comment :)

The Omniglot experiment seems to be working. I just had to let it run a bit longer to get similar results:

INFO:root:episode batch 740, avg train loss 0.100891
INFO:root:validation overall accuracy 0.768667
INFO:root:0-shot: 0.076, 1-shot: 0.852, 2-shot: 0.888, 3-shot: 0.932, 4-shot: 0.900, 5-shot: 0.964,

**Comment by lukaszkaiser** [comment]
This is great, thanks! Will you make a PR with the correction? That'd be great :).

**Comment by kasrahbar** [comment]
Thanks Lukasz! Here's the PR: https://github.com/kasrahbar/models/pull/1

**Comment by lukaszkaiser** [comment]
I didn't notice this, but the PR above is in your private repo - you need to push it here! It's a great PR but can you do it so we can merge it here? (It must be a PR against this repo.)

**Comment by kasrahbar** [comment]
Oops, here's the PR again: https://github.com/tensorflow/models/pull/5566
